new-log-viewer/src/services/decoders/JsonlDecoder.ts

```diff
@@ -0,0 +1,240 @@
+        }
+        const dayjsObj: dayjs.Dayjs = dayjs.utc(timestamp);
+        const formattedDate = dayjsObj.format(this.#dateFormat);
+        input = input.replace(this.#datePattern, formattedDate);
+
+        return [
+            dayjsObj.valueOf(),
+            input,
+        ];
+    }
+
+    /**
+     * Replaces placeholders in the input string with corresponding properties from the logEvent
+     * object.
+     *
+     * @param input The input string with placeholders to be replaced.
+     * @param logEvent The log event object containing properties to replace the placeholders.
+     * @return - The input string with placeholders replaced by corresponding property values.
+     * @private
+     */
+    #formatVariables (input: string, logEvent: JsonObject) {
+        // Replace each placeholder with the corresponding property from logEvent
+        for (const propName of this.#propertyNames) {
+            if (propName in logEvent) {
+                const placeholder = `%${propName}`;
+                const propValue = logEvent[propName];
+                let propValueString: string;
+                if ("undefined" === typeof propValue || null === propValue) {
+                    console.error(`Unexpected undefined logEvent[${propName}]`);
+                    propValueString = "undefined";
+                } else if ("object" === typeof propValue) {
+                    propValueString = JSON.stringify(propValue);
+                } else {
+                    propValueString = propValue.toString();
+                }
+                input = input.replace(placeholder, propValueString);
+            }
+        }
+
+        input += "\n";
+
+        return input;
+    }
+
+    /**
+     * Extracts the verbosity from the given log event.
+     *
+     * @param logEvent The log event containing the verbosity.
+     * @return The extracted verbosity.
+     */
+    #extractVerbosity (logEvent: JsonObject) {
+        let verbosity = LOG_VERBOSITY.NONE;
+        const verbosityString: string = logEvent[this.#verbosityPropName] as string;
+        if (false === (verbosityString in LOG_VERBOSITY)) {
+            console.error(`Unable to find verbosity from key ${this.#verbosityPropName}` +
+                ` of type ${typeof verbosityString}`);
+        } else {
+            verbosity = LOG_VERBOSITY[verbosityString as (keyof typeof LOG_VERBOSITY)];
+        }
+
+        return verbosity;
+    }
+
+    constructor (dataArray: Uint8Array) {
+        this.#dataArray = dataArray;
+    }
+
+    setDecodeOptions (options: JsonlDecodeOptionsType): boolean {
+        this.#setTextPattern(
+            options.textPattern
+        );
+        this.#timestampPropName = options.timestampPropName;
+        this.#verbosityPropName = options.verbosityPropName;
+
+        return true;
+    }
+
+    buildIdx (): number {
+        const text = JsonlDecoder.TEXT_DECODER.decode(this.#dataArray);
+        const split = text.split("\n");
```

Comment:
Mind elaborating? The `split()` JS API has long been known for its efficiency due to V8 optimizations, but I guess not in all cases especially when we have a large string.

Reply:
hmm... Mind showing your benchmarking code? With below `buildIdx ()` constantly takes 1400-1500ms while the original code takes 1300-1400ms.

```
    buildIdx (): number {
        const text = `${JsonlDecoder.#textDecoder.decode(this.#dataArray)}\n`;
        let startIndex = 0;
        let endIndex = text.indexOf("\n");

        while (-1 !== endIndex) {
            const line = text.slice(startIndex, endIndex);
            if (0 < line.length) {
                try {
                    const logEvent = JSON.parse(line) as JsonObject;
                    this.#logEvents.push(logEvent);
                } catch (e) {
                    console.error(e, line);
                }
            }
            startIndex = endIndex + 1;
            endIndex = text.indexOf("\n", startIndex);
        }

        return this.#logEvents.length;
    }
```

Reply:
```js
    buildIdx (): number {
        const text = JsonlDecoder.#textDecoder.decode(this.#dataArray);

        const beginTimestamp = performance.now();

        const split = text.split("\n");
        for (const line of split) {
            if (0 === line.length) {
                continue;
            }

            try {
                const logEvent = JSON.parse(line) as JsonObject;
                this.#logEvents.push(logEvent);
            } catch (e) {
                console.error(e, line);
            }
        }

        // let beginIdx = 0;
        // while (beginIdx < text.length) {
        //     const endIdx = text.indexOf("\n", beginIdx);
        //     let line;
        //     if (-1 === endIdx) {
        //         line = text.substring(beginIdx);
        //         beginIdx = text.length;
        //     } else {
        //         line = text.substring(beginIdx, endIdx);
        //         beginIdx = endIdx + 1;
        //     }
        //     if (0 === line.length) {
        //         continue;
        //     }
        //
        //     try {
        //         const logEvent = JSON.parse(line) as JsonObject;
        //         this.#logEvents.push(logEvent);
        //     } catch (e) {
        //         console.error(e, line);
        //     }
        // }

        const endTimestamp = performance.now();
        console.log(endTimestamp - beginTimestamp, "ms");

        return this.#logEvents.length;
    }
```

Compared the two implementations above on a 94MB log file.

Reply:
Your `indexOf` implementation is indeed slightly faster than mine. When I was testing with the 512MB log file, the speed up compared the my original `split` implementation was about 30ms with the `performance.now()` counters. 

However, after I removed the `performance.now()` counters and used the browser profiler (Debugger - Performance tab) to measure the time, both your `indexOf` implementation and my `split` seem to take 1.32 ~ 1.33 s consistently. I did notice the JS heap size is 8MB less (468MB v.s. 476MB) with your `indexOf`, which makes sense as we previously suspect `split` could be caching chunks of split results to speed up the accesses just in time.

I imagine the results could be vastly different if we have some very long lines in the test files.

With that said, both implementations are concise and should be easy to maintain. Which you do you prefer?

Reply:
In both implementations, I propose we can move the `if (0 === line.length)` check in the `catch` clause since it's uncommon to have empty lines in the middle of such log files. The performance speed up with the 512M file was about 5ms. e.g.,
```
            try {
                const logEvent = JSON.parse(line) as JsonObject;
                this.#logEvents.push(logEvent);
            } catch (e) {
                if (0 === line.length) {
                    console.error(e, line);
                }
            }
```